### PR TITLE
monitoring: add observability.silenceAlerts

### LINKED ISSUE #11210: explore solutions for "impossible to silence monitoring alerts"
Sourcegraph.com has some alerts always firing (e.g. gitserver disk space) that we do not care about on Sourcegraph.com. Customers will likely have a few outliers like this, too, where silencing will be important.

We should have a mechanism for silencing these _somehow_.

### PR DESCRIPTION
**Pull request by bobheadxi**
Adds `observability.silenceAlerts`, which deploys silences to `sourcegraph/prometheus`'s built-in Alertmanager.

**updated**

```json
{
  "observability.silenceAlerts": [
    "warning_gitserver_disk_space_remaining",
    "critical_gitserver_command_duration_test"
  ]
}
```

Closes #11210

### Considerations

* Specificity: I'm opting to disallow broad silences (ie all of `name`, `level`, `service` are required and a user cannot mute all `"service": "gitserver"` alerts). Regex is _possible_, but not allowed at the moment. Open to thoughts on this though
* Implementation detail: Alertmanager requires a start and end date for each alert. Right now, I've set this to 10 years... should be fine right?
* Via a normal user flow of:
    1. see site banner for critical alert
    2. go to grafana

   there does not seem to be a clear way to get the `name` of an alert at the moment - not sure if this is within the scope of this PR to address, and if there's a good way to go about it (maybe replace the human-readable description in the main table with alert name?) => resolved via generating full alert names in docs

### CODE REVIEW

**Comment by slimsag** [comment]
> Specificity: I'm opting to disallow broad silences (ie all of name, level, service are required and a user cannot mute all "service": "gitserver" alerts). Regex is possible, but not allowed at the moment. Open to thoughts on this though

This sounds good to me, bulk silencing seems like a really bad practice in general so I'd rather not have this and find out we need it later - than enable someone to shoot themselves in the foot so-to-speak :) 

> Implementation detail: Alertmanager requires a start and end date for each alert. Right now, I've set this to 10 years... should be fine right?

Seems reasonable to me, I was just thinking that it is quite cool that with this approach we can un-silence alerts or notify admins in the future to un-silence them if we e.g. see they are flaky on many instances and fix them in an upgrade. That's a cool property of having this in site config :)

> there does not seem to be a clear way to get the name of an alert at the moment - not sure if this is within the scope of this PR to address, and if there's a good way to go about it (maybe replace the human-readable description in the main table with alert name?)

This is indeed unfortunate.. I really like having the human-readable names, though. What about having admins first go to the solutions doc (as they would to consider if it might be a real issue for them), and having one of the pre-generated solutions there being "Silence the alert in your site config: `<name>`"?

This would also mean you could have the site config be this format instead:

```json
{
  "observability.silenceAlerts": [
    "warning_gitserver_disk_space_remaining",
    "critical_gitserver_command_duration_test"
  ]
}
```

WDYT?

File: cmd/frontend/internal/app/debug_test.go [hunk 1]
```diff
@@ -49,7 +49,7 @@ func Test_prometheusValidator(t *testing.T) {
 			wantProblemSubstring: "",
 		},
 		{
-			name: "prometheus not found",
+			name: "prometheus not found (with obs. alerts)",
```
Comment (slimsag) [inline cmd/frontend/internal/app/debug_test.go:52]:
```suggestion
			name: "prometheus not found (without obs. alerts)",
```

Intentional or not?

File: docker-images/prometheus/cmd/prom-wrapper/change.go [hunk 2]
```diff
@@ -102,3 +108,86 @@ func changeSMTP(ctx context.Context, log log15.Logger, change ChangeContext, new
 
 	return
 }
+
+func changeSilences(ctx context.Context, log log15.Logger, change ChangeContext, newConfig *subscribedSiteConfig) (result ChangeResult) {
```
Comment (slimsag) [inline docker-images/prometheus/cmd/prom-wrapper/change.go:112]: Add docstring about when this called and what it handles at a high-level.

File: docker-images/prometheus/cmd/prom-wrapper/change.go [hunk 3]
```diff
@@ -102,3 +108,86 @@ func changeSMTP(ctx context.Context, log log15.Logger, change ChangeContext, new
...
+	// convenience function for creating a prefixed problem - this reflects the relevant site configuration fields
+	newProblem := func(err error) {
+		result.Problems = append(result.Problems, conf.NewSiteProblem(fmt.Sprintf("`observability.silenceAlerts`: %v", err)))
+	}
+
+	var (
+		createdBy      = "src-prom-wrapper"
+		comment        = "Applied via `observability.silenceAlerts` in site configuration"
+		startTime      = strfmt.DateTime(time.Now())
+		endTime        = strfmt.DateTime(time.Now().Add(10 * 365 * 24 * time.Hour)) // 10 year expiry - TODO: do we need to worry about this?
```
Comment (slimsag) [inline docker-images/prometheus/cmd/prom-wrapper/change.go:122]:
```suggestion
        // 10 year expiry, we don't want it to expire and will remove the silence once it is removed from site config
		endTime        = strfmt.DateTime(time.Now().Add(10 * 365 * 24 * time.Hour))
```

**Comment by bobheadxi** [comment]
> This is indeed unfortunate.. I really like having the human-readable names, though. What about having admins first go to the solutions doc (as they would to consider if it might be a real issue for them), and having one of the pre-generated solutions there being "Silence the alert in your site config: <name>"?

> This would also mean you could have the site config be this format instead

Sounds good!

File: cmd/frontend/internal/app/debug_test.go (hunk 1 above)
Comment (bobheadxi) [reply cmd/frontend/internal/app/debug_test.go]: This is intentional - this pair of `prometheus not found` tests just checks if you have observability alerts or observability silence alerts

tried to clarify in 1799e033bbd99886bbe5d53417d6a9be74b995e3

**Comment by daxmc99** [comment]
Looks good! 
Just wanted to confirm that 
```json
{
  "observability.silenceAlerts": [
    "warning_gitserver_disk_space_remaining",
    "critical_gitserver_command_duration_test"
  ]
}
```

in the site config will be the typical way of enabling this.

**Comment by bobheadxi** [comment]
@daxmc99 yep, should be unless anyone has any other thoughts on that! Working on an update to this PR to enable that, but it shouldn't be a significant change implementation-wise

File: docker-images/prometheus/cmd/prom-wrapper/change.go (hunk 2 above)
Comment (bobheadxi) [reply docker-images/prometheus/cmd/prom-wrapper/change.go:113]: a237d18

**Comment by bobheadxi** [comment]
2c3f7b9 , e5fa2a0 makes the change to specify silences by alertname and generates alertname in the alert solutions documentaiton

**Comment by bobheadxi** [comment]
image

@slimsag let me know if there's any follow-ups you want me to make!

**Pull request merged**